Firestore/core/src/firebase/firestore/remote/remote_store.mm

```diff
@@ -291,6 +289,187 @@
+      mutationResults:std::move(mutation_results)
+          streamToken:write_stream_->GetLastStreamToken()];
+  [sync_engine_ applySuccessfulWriteWithResult:batchResult];
+
+  // It's possible that with the completion of this mutation another slot has
+  // freed up.
+  FillWritePipeline();
+}
+
+/**
+ * Handles the closing of the StreamingWrite RPC, either because of an error or
+ * because the RPC has been terminated by the client or the server.
+ */
+void RemoteStore::OnWriteStreamClose(const Status& status) {
+  if (status.ok()) {
+    // Graceful stop (due to Stop() or idle timeout). Make sure that's
+    // desirable.
+    HARD_ASSERT(!ShouldStartWriteStream(),
+                "Write stream was stopped gracefully while still needed.");
+  }
+
+  // If the write stream closed due to an error, invoke the error callbacks if
+  // there are pending writes.
+  if (!status.ok() && !write_pipeline_.empty()) {
+    if (write_stream_->handshake_complete()) {
+      // This error affects the actual writes.
+      HandleWriteError(status);
+    } else {
+      // If there was an error before the handshake finished, it's possible that
+      // the server is unable to process the stream token we're sending.
+      // (Perhaps it's too old?)
+      HandleHandshakeError(status);
+    }
+  }
+
+  // The write stream might have been started by refilling the write pipeline
+  // for failed writes
+  if (ShouldStartWriteStream()) {
+    StartWriteStream();
+  }
+}
+
+void RemoteStore::HandleHandshakeError(const Status& status) {
+  HARD_ASSERT(!status.ok(), "Handling write error with status OK.");
+
+  // Reset the token if it's a permanent error, signaling the write stream is
+  // no longer valid. Note that the handshake does not count as a write: see
+  // comments on `Datastore::IsPermanentWriteError` for details.
+  if (Datastore::IsPermanentError(status)) {
+    NSString* token =
+        [write_stream_->GetLastStreamToken() base64EncodedStringWithOptions:0];
+    LOG_DEBUG("RemoteStore %s error before completed handshake; resetting "
+              "stream token %s: "
+              "error code: '%s', details: '%s'",
+              this, token, status.code(), status.error_message());
+    write_stream_->SetLastStreamToken(nil);
+    [local_store_ setLastStreamToken:nil];
+  } else {
+    // Some other error, don't reset stream token. Our stream logic will just
+    // retry with exponential backoff.
+  }
+}
+
+void RemoteStore::HandleWriteError(const Status& status) {
+  HARD_ASSERT(!status.ok(), "Handling write error with status OK.");
+
+  // Only handle permanent errors here. If it's transient, just let the retry
+  // logic kick in.
+  if (!Datastore::IsPermanentWriteError(status)) {
+    return;
+  }
+
+  // If this was a permanent error, the request itself was the problem so it's
+  // not going to succeed if we resend it.
+  FSTMutationBatch* batch = write_pipeline_.front();
+  write_pipeline_.erase(write_pipeline_.begin());
```

Comment:
`std::queue` cannot be used because of the `for` loop at line 351. My impression is that `deque` starts outperforming `vector` only for a very large dataset (something like tens of thousands of elements). All things considered, I'm slightly in favor of the more common `vector`.